Change design of the coupon block [MAILPOET-4890]

lysyjan opened the pull request:
## Description

This PR primarily refactors the coupon block settings to React because it simplifies designing the settings panel.
Also, it was necessary to extend or add some functionality.

## Code review notes

I would recommend reviewing the PR by commits.
Unfortunately, I had to remove one of the test cases for coupon block in newsletter editor tests.

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4890]

## After-merge notes

Don't forget to extend the ticket with test improvements.


[MAILPOET-4890]: https://mailpoet.atlassian.net/browse/MAILPOET-4890?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

lysyjan committed:
Move margin to lower level

[MAILPOET-4890]

lysyjan committed:
Use React for coupon source switch

[MAILPOET-4890]

lysyjan committed:
Use React for coupon general settings

[MAILPOET-4890]

lysyjan committed:
Add coupon settings component

[MAILPOET-4890]

lysyjan committed:
Make general panel switchable

[MAILPOET-4890]

lysyjan committed:
Rename component

[MAILPOET-4890]

lysyjan committed:
Create usage restriction component

[MAILPOET-4890]

lysyjan committed:
Move products selections to React component

[MAILPOET-4890]

lysyjan committed:
Move category selections to React component

[MAILPOET-4890]

lysyjan committed:
Move email restrictions to React component

[MAILPOET-4890]

lysyjan committed:
Add validation for email restrictions

[MAILPOET-4890]

lysyjan committed:
Create component for usage limits

[MAILPOET-4890]

lysyjan committed:
Create new component for existing coupons

[MAILPOET-4890]

lysyjan committed:
Add reseting code when tab is changed

[MAILPOET-4890]

lysyjan committed:
Make display options hideable

[MAILPOET-4890]

lysyjan committed:
Unify design with other blocks

[MAILPOET-4890]

lysyjan committed:
Add validation improvements to coupon settings

[MAILPOET-4890]

lysyjan committed:
Fix renamed coupon block attributes

[MAILPOET-4890]

lysyjan committed:
Remove settings coupon block test

Because the new settings panel was refactored to React. It is not possible to use this test anymore.
[MAILPOET-4890]

triple0t commented on mailpoet/lib/AdminPages/Pages/NewsletterEditor.php:
```diff
@@ -136,6 +144,8 @@ public function render() {
       'is_wc_transactional_email' => $newsletterId === $woocommerceTemplateId,
       'is_confirmation_email_template' => $newsletterId === $confirmationEmailTemplateId,
       'is_confirmation_email_customizer_enabled' => (bool)$this->settings->get('signup_confirmation.use_mailpoet_editor', false),
+      'product_categories' => $this->wpPostListLoader->getWooCommerceCategories(),
+      'products' => $this->wpPostListLoader->getProducts(),
```
Suggestion: Why not move these two into the [woocommerceData](https://github.com/mailpoet/mailpoet/blob/d06447b48bff6ec6a2c29088c9e2ab91030e258f/mailpoet/lib/AdminPages/Pages/NewsletterEditor.php#L112) section?

This would prevent any error that product_categories & products could cause. e.g, when WooCommerce is not available

lysyjan replied:
I have checked the code, and if I am not wrong, the result is empty when WC is not installed. So it looks safe to me.

Product and categories are here because the selection component reads them by the property `endpoint`. I used this approach because is also used in the dynamic segments form.

lysyjan replied:
> Can this component take all the products on a site?
I noticed you removed the pagination. This means this component will contain all the products available on the site.

I will continue this topic here because it is related. I am aware of it, and we use this approach in dynamic segments.
But I agree it is not the best solution for sites with a higher amount of products. I wanted to change my solution and use the property `remoteQuery`([#](https://github.com/mailpoet/mailpoet/blob/d116ff360b8d73f0ae09429ae0c365ed30a54def/mailpoet/assets/js/src/form/fields/selection.jsx#L150-L182)). Unfortunately, I cannot map the response from the endpoint `automatedLatestContent` without significant changes as it is [here](https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/assets/js/src/newsletter_editor/blocks/coupon.ts#L434-L447).

I prefer to create a new ticket because this PR covers a lot of other changes. The previous solution contained this benefit, but I don't want a huge PR and want to unblock the rest of the tasks. Do you agree?

triple0t replied:
Yeah, that is fine with me.

triple0t commented on mailpoet/assets/js/src/newsletter_editor/blocks/coupon/usage_restriction.tsx:
```diff
@@ -0,0 +1,275 @@
+};
+
+type Props = {
+  getValueCallback: (name: string) => string | boolean | Backbone.Collection;
+  setValueCallback: (
+    name: string,
+    value: string | boolean | Backbone.Collection,
+  ) => void;
```
I noticed we repeat this Type declaration in a few other files. I would suggest creating a reusable Type that you can import and use everywhere else.

lysyjan replied:
I noticed it during the implementation, but in the end, I forgot. I will change it.

triple0t commented on mailpoet/assets/js/src/newsletter_editor/blocks/coupon/usage_limits.tsx:
```diff
@@ -0,0 +1,80 @@
+  setValueCallback: (
+    name: string,
+    value: string | boolean | Backbone.Collection,
+  ) => void;
```
Same here as well

triple0t commented on mailpoet/assets/js/src/newsletter_editor/blocks/coupon/usage_limits.tsx:
```diff
@@ -0,0 +1,80 @@
+        >
+          <PanelRow>
+            <TextControl
+              className="mailpoet_field_coupon_usage_limits_per_coupon"
```
Does this `className` serve any purpose here?
If not, May I suggest removing it?

lysyjan replied:
I added those classes automatically because I didn't know if I needed them in one of my following improvements, such as design or validation. I can remove them.

triple0t commented on mailpoet/assets/js/src/newsletter_editor/blocks/coupon/usage_limits.tsx:
```diff
@@ -0,0 +1,80 @@
+          </PanelRow>
+          <PanelRow>
+            <TextControl
+              className="mailpoet_field_coupon_usage_limits_per_user"
```
Same here

triple0t commented on mailpoet/assets/js/src/newsletter_editor/blocks/coupon/usage_restriction.tsx:
```diff
@@ -0,0 +1,275 @@
+          </PanelRow>
+          <PanelRow>
+            <label htmlFor="productIds">{MailPoet.I18n.t('products')}</label>
+            <Selection
+              field={{
+                ...productsField,
+                name: 'productIds',
+                selected: () =>
+                  this.state.productIds.map((product: Post) => product.id),
+              }}
+              onValueChange={this.handleSelection}
+            />
+          </PanelRow>
```
Can this component take all the products on a site?
I noticed you removed the pagination. This means this component will contain all the products available on the site.

triple0t reviewed (commented):
Such an incredible amount of work. Well done, @lysyjan.

I have a few comments.

1. EmailRestrictions does not allow multiple emails
   <img width="360" alt="Screenshot 2023-02-16 at 1 24 40 PM - emailRestrictions" src="https://user-images.githubusercontent.com/30554163/219372297-436b92c2-28a2-42a3-a8d6-3bf5af8d80d9.png">

   <img width="664" alt="Screenshot 2023-02-16 at 2 01 15 PM" src="https://user-images.githubusercontent.com/30554163/219372352-a7bad166-1166-4e8b-ac8f-6985d0e29093.png">

   As you may notice, we are unable to save multiple emails in the `emailRestrictions` field. This should be allowed.

2. Product ids and Categories id not saving.
I assume you have to update these [methods](https://github.com/mailpoet/mailpoet/blob/d06447b48bff6ec6a2c29088c9e2ab91030e258f/mailpoet/lib/WooCommerce/CouponPreProcessor.php#L121-L127)
   <img width="437" alt="Screenshot 2023-02-16 at 1 37 12 PM" src="https://user-images.githubusercontent.com/30554163/219372920-579c9e75-3f96-44aa-bde1-62c18cb7d498.png">

3. Error in the console
   <img width="1792" alt="Screenshot 2023-02-16 at 1 27 31 PM - console" src="https://user-images.githubusercontent.com/30554163/219373348-7bb83fa0-c5f3-42fc-9d92-982d2e1ae4b2.png">
   I got this error when I tried editing some input fields. e.g, Minimum spend

4. Kind reminder to create a follow-up ticket for the tests removed in this [commit](https://github.com/mailpoet/mailpoet/pull/4733/commits/d06447b48bff6ec6a2c29088c9e2ab91030e258f)

5. "Display options" button/link is really easy to miss. Are we sure we want to keep it hidden by default? 

   <img width="357" alt="Screenshot 2023-02-16 at 1 14 02 PM - Display options" src="https://user-images.githubusercontent.com/30554163/219373855-f51b8553-b998-494e-a900-882870624ca5.png">

6. Can we duplicate the "Back to selection" button/link? I suggest we add to the selection panel's top and bottom. It's really easy to miss as well

   <img width="359" alt="Screenshot 2023-02-16 at 1 16 13 PM - Back to selection" src="https://user-images.githubusercontent.com/30554163/219374086-a7cc41a5-d810-402e-aeb7-58549341040b.png">

lysyjan commented:
@triple0t Thank you for the detailed review with nice catches. Those are big changes that I wanted to move forward with.
I want to address some of your findings from the list.

> 1. EmailRestrictions does not allow multiple emails

I tried it again on my local, and I think the space isn't allowed behind the comma. In my opinion, it's fine. I also tested the behavior in the WC plugin, and they allow the spaces, but after saving are removed. This would mean more work for this PR, and if you think that is important, I prefer addressing it in another ticket.
![Email restirctions](https://d.pr/i/cxTRad+)

> 2. Product ids and Categories id not saving.

Nice catch. I will fix it.

> 3. Error in the console

This can happen when the initial value is `null` or `undefined`. I will fix it. Thanks.

> 4. Kind reminder to create a follow-up ticket for the tests removed in this [commit](https://github.com/mailpoet/mailpoet/pull/4733/commits/d06447b48bff6ec6a2c29088c9e2ab91030e258f)

I'm still thinking about those tests. We have an existing ticket for test improvements. I will add it to after merge notes.

> 5. "Display options" button/link is really easy to miss. Are we sure we want to keep it hidden by default?

This is based on the design in Figma. I made it a little bit different and followed our current approach in the other blocks.

> 6. Can we duplicate the "Back to selection" button/link? I suggest we add to the selection panel's top and bottom. It's really easy to miss as well

Yes, we can. But I would expect some spacing or separator here.

@pavel-mailpoet Please, can you tell us your opinion on points 5 and 6?
For me, those are details that can be decided and improved later.

pavel-mailpoet commented:
@lysyjan 

5. "Display options" is consistent with ALC and seems fine to me so I would keep it as you did.

6. I agree with you. If it is easy, you can duplicate the button, but please add some margin so it looks good.


Btw. nice animation you did when switching to and from the display option. If this was not the legacy editor I would probably create a new ticket to add more animations when switching between other tabs, but it is not worth for the old editor.

lysyjan commented:
>Btw. nice animation you did when switching to and from the display option. If this was not the legacy editor I would probably create a new ticket to add more animations when switching between other tabs, but it is not worth for the old editor.

@pavel-mailpoet It's reused style from the other blocks.

lysyjan committed:
Fix mapping products and categories for new coupon

[MAILPOET-4890]

lysyjan committed:
Add default values

Because React components throw errors when the initial value is null or undefined before the change.
I added default values for all inputs.
[MAILPOET-4890]

lysyjan committed:
Create common types for callbacks

[MAILPOET-4890]

lysyjan committed:
Remove redundant classes

[MAILPOET-4890]

lysyjan committed:
Fix spacing in block settings

After the initial change for the new coupon block was necessary to fix overlooked issues.
[MAILPOET-4890]

lysyjan commented:
@triple0t I tried to address all your remarks.
Unfortunately, when I tried to add `Back to selection link`, it didn't look good, in my opinion. I would prefer a tip from a designer here. I guess that when we prepare a demo we will have time for those details.

triple0t reviewed (approved):
LGTM.
Well done 👏

Aschepikov reviewed (approved):
GJ!
